File: src/main/java/no/nav/pto/veilarbfilter/client/VeilarbveilederClient.java
package no.nav.pto.veilarbfilter.client;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import no.nav.common.rest.client.RestClient;
import no.nav.common.rest.client.RestUtils;
import no.nav.common.token_client.client.AzureAdMachineToMachineTokenClient;
import no.nav.common.types.identer.EnhetId;
import no.nav.common.utils.EnvironmentUtils;
import no.nav.pto.veilarbfilter.config.EnvironmentProperties;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

import static no.nav.common.client.utils.CacheUtils.tryCacheFirst;
import static no.nav.common.utils.UrlUtils.joinPaths;
import static org.springframework.http.HttpHeaders.AUTHORIZATION;

@Slf4j
@Service
public class VeilarbveilederClient {
    private final String veilarbveilederBaseUrl;
    private final OkHttpClient client;
    private final Cache<EnhetId, List<String>> hentVeilederePaaEnhetCache;
    private final Supplier<String> systemUserTokenProvider;

    @Autowired
    public VeilarbveilederClient(EnvironmentProperties properties, AzureAdMachineToMachineTokenClient tokenClient) {
        final String appName = "veilarbveileder";
        final String namespace = "pto";
        this.veilarbveilederBaseUrl = properties.getVeilarbveilederUrl() + "/veilarbveileder";
        this.client = RestClient.baseClient();
        systemUserTokenProvider = () ->

                tokenClient.createMachineToMachineToken(String.format("api://%s-fss.%s.%s/.default",
                        (EnvironmentUtils.isProduction().orElseThrow()) ? "prod" : "dev", namespace, appName)
                );

        hentVeilederePaaEnhetCache = Caffeine.newBuilder()
                .expireAfterWrite(5, TimeUnit.MINUTES)
                .maximumSize(600)
                .build();
    }

    public List<String> hentVeilederePaaEnhet(EnhetId enhet) {
        return tryCacheFirst(hentVeilederePaaEnhetCache, enhet,
                () -> hentVeilederePaaEnhetQuery(enhet));
    }

    @SneakyThrows
    private List<String> hentVeilederePaaEnhetQuery(EnhetId enhet) {
        Request request = new Request.Builder()
                .header(AUTHORIZATION, "Bearer " + systemUserTokenProvider.get())
                .url(joinPaths(veilarbveilederBaseUrl, "/api/enhet/", enhet.get(), "/identer"))
                .build();

        try (Response response = client.newCall(request).execute()) {
            RestUtils.throwIfNotSuccessful(response);
            return RestUtils.parseJsonResponseArrayOrThrow(response, String.class);
        }
    }
}

File: src/main/kotlin/no/nav/pto/veilarbfilter/client/VeilarbveilederClient.kt
package no.nav.pto.veilarbfilter.client

import io.ktor.client.HttpClient
import io.ktor.client.engine.apache.Apache
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.response.HttpResponse
import io.ktor.client.response.readText
import io.ktor.http.HttpHeaders
import kotlinx.coroutines.runBlocking
import no.nav.pto.veilarbfilter.ObjectMapperProvider
import no.nav.pto.veilarbfilter.config.Configuration


data class Enhet (val enhetId: String, val navn: String)
data class Veileder(val etternavn: String?, val fornavn: String?, val ident: String, val navn: String?)

data class VeiledereResponse (val enhet: Enhet, val veilederListe: List<Veileder>)


private val veilederPaEnhetenCache = VeilederCache()

class VeilarbveilederClient (config: Configuration) {
    private val veilarbveilederClientUrl = config.veilarbveilederConfig.url

    fun hentVeilederePaEnheten(enhetId: String, requestToken: String?): VeiledereResponse {
        requireNotNull(requestToken) { "RequestToken is not set" }
        val veilederCacheValue = veilederPaEnhetenCache.veilederePaEnheten(enhetId)

        if(veilederCacheValue !== null) {
            return veilederCacheValue
        }

        return runBlocking {
            HttpClient(Apache).use { httpClient ->
                val result = httpClient.get<HttpResponse>("$veilarbveilederClientUrl/enhet/$enhetId/veiledere") {
                    header(HttpHeaders.Authorization, "Bearer $requestToken")
                }
                if (result.status.value != 200) {
                    throw RuntimeException("Veilarbveileder kallet feilet ${result.status.description}")
                }
                val res = result.readText()
                val veiledereResponse = ObjectMapperProvider.objectMapper.readValue(res, VeiledereResponse::class.java)
                veilederPaEnhetenCache.leggTilEnhetICachen(enhetId, veiledereResponse)
                veiledereResponse
            }
        }
    }

}